### PR TITLE
feat: support message key as a column in avro_upsert

### PR DESCRIPTION
**Pull request by tabVersion**
Signed-off-by: tabVersion <tabvision@bupt.icu>I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## Checklist For Contributors

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [ ] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

**Review by hzxa21 — COMMENTED** [review]
Let's fill-in the Documentation section in the PR description to include some example usages and basic documentation. Also, what tests are done for this PR?

File: src/connector/src/parser/avro/parser.rs [hunk 1]
```diff
@@ -62,6 +66,7 @@ impl AvroParserConfig {
         use_schema_registry: bool,
         enable_upsert: bool,
         upsert_primary_key_column_name: Option<String>,
+        key_as_column: bool,
```
Comment (idx0-dev) [inline src/connector/src/parser/avro/parser.rs:69]: Does `key_as_column` require `enable_upsert` to be enabled?

File: src/connector/src/parser/avro/parser.rs [hunk 2]
```diff
@@ -184,34 +225,63 @@ impl AvroParser {
...
+                let (schema_id, mut raw_payload) = extract_schema_id(&key_payload)?;
+                let writer_schema = resolver.get(schema_id).await?;
+                let key_avro_value = from_avro_datum(
+                    writer_schema.as_ref(),
+                    &mut raw_payload,
+                    self.key_schema.as_deref(),
+                ).map_err(|e| RwError::from(ProtocolError(e.to_string())))?;
+
+                if let Value::Record(fields) = value {
+                    let mut fields = fields;
```
Comment (idx0-dev) [inline src/connector/src/parser/avro/parser.rs:274]:
```suggestion
                if let Value::Record(mut fields) = value {
```

**Comment by idx0-dev** [comment]
Maybe we should check the hard-coded `key_field` field is not in the value schema?

File: src/connector/src/parser/avro/parser.rs [hunk 3]
```diff
@@ -184,34 +225,63 @@ impl AvroParser {
...
-                (msg.record, Op::Insert)
-            } else {
-                (msg.primary_key, Op::Delete)
+            match (!msg.record.is_empty(), self.key_as_column) {
+                (true, false) => (None, msg.record, Op::Insert),
+                (true, true) => (Some(msg.primary_key), msg.record, Op::Insert),
+                (false, false) => (None, msg.primary_key, Op::Delete),
+                (false, true) => {
+                    return Err(RwError::from(ProtocolError(format!(
+                        "key_as_column is not supported when payload is empty, key: {:?}",
```
Comment (idx0-dev) [inline src/connector/src/parser/avro/parser.rs:241]: We should emit an `Op::Delete` here to fullfill the upsert semantic.

**Comment by tabVersion** [comment]
feature has been done in #9926

**Pull request closed without merging**